Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp

```diff
@@ -0,0 +1,415 @@
+{
+    UNUSED_PARAM(pointerType);
+
+    auto location = deviceScaleLocationInView(page, locationInView);
+    unsigned wpeButton = mouseButtonToWPEButton(button);
+    auto modifier = stateModifierForWPEButton(wpeButton);
+    uint32_t state = modifiersToEventState(keyModifiers) | currentModifiers;
+
+    switch (interaction) {
+    case MouseInteraction::Move:
+        doMotionEvent(page, location, state);
+        break;
+    case MouseInteraction::Down:
+        currentModifiers |= modifier;
+        doMouseEvent(page, location, wpeButton, 1, state | modifier);
+        break;
+    case MouseInteraction::Up:
+        currentModifiers &= ~modifier;
+        doMouseEvent(page, location, wpeButton, 0, state & ~modifier);
+        break;
+    case MouseInteraction::SingleClick:
+        doMouseEvent(page, location, wpeButton, 1, state | modifier);
+        doMouseEvent(page, location, wpeButton, 0, state);
+        break;
+    case MouseInteraction::DoubleClick:
+        doMouseEvent(page, location, wpeButton, 1, state | modifier);
+        doMouseEvent(page, location, wpeButton, 0, state);
+        doMouseEvent(page, location, wpeButton, 1, state | modifier);
+        doMouseEvent(page, location, wpeButton, 0, state);
+        break;
+    }
+    return;
+}
+#endif // ENABLE(WEBDRIVER_MOUSE_INTERACTIONS)
+
+#if ENABLE(WEBDRIVER_KEYBOARD_INTERACTIONS)
+static void doKeyStrokeEvent(WebPageProxy &page, bool pressed, uint32_t keyCode, uint32_t modifiers, bool doReleaseAfterPress = false)
+{
+    auto* view = page.wpeView();
+    auto* display = wpe_view_get_display(view);
+    if (!display) {
+        LOG(Automation, "WebAutomationSession::doKeyStrokeEvent: Failed to find WPE display. Ignoring event.");
+        return;
+    }
+    GUniqueOutPtr<GError> error;
+    auto* keymap = WPE_KEYMAP(wpe_display_get_keymap(display, &error.outPtr()));
+    if (error) {
+        LOG(Automation, "WebAutomationSession::doKeyStrokeEvent: Failed to get keymap: %s. Ignoring event.", error->message);
+        return;
+    }
+
+    WPEKeymapEntry* entries;
+    guint entriesCount;
+    if (!wpe_keymap_get_entries_for_keyval(keymap, keyCode, &entries, &entriesCount)) {
+        LOG(Automation, "WebAutomationSession::doKeyStrokeEvent: Failed to get keymap entries for keyval %u. Ignoring event.", keyCode);
+        return;
+    }
+    unsigned keyval = entries[0].keycode;
+
+    // Note: wpe_event_keyboard_new uses the name 'keycode' for the value with the hardware key code
+    // and 'keyval' for the value with the translated value of the key (e.g. ascii-like).
+    // In comparison to the usage of this function in Wayland callbacks, here we have the variable names swapped,
+    // with WebAutomationSession using `keyCode` for the translated value and `keyval` for the XKB key code.
```

Comment:
This is not what we want. First let's rename variables to avoid confusion, using always keyvale for the key symbol and keycode for the hardware code. What this function receives is a keyval, that we need to translate, so the first thing we need is to get the keycode associated the keyval using wpe_keymap_get_entries_for_keyval() as you are doing. Then we need to translate the keycode + modifiers into the actual keyval (so for example a + Shift ends up being A). For that we need new API in WPE platform, see https://github.com/WebKit/WebKit/pull/28524

Reply:
Done. Indeed, now if we send `key_down(SHIFT).key_down('e').perform()` we get an `E` and no modifiers instead of `e` + modifier.